source/llassetgen/source/Image.cpp

```diff
@@ -0,0 +1,358 @@
+#include <vector>
+#include <iostream>
+#include <fstream>
+#include <cassert>
+#include <cmath>
+
+/*
+ * If we sort png.h below the freetype includes, an compile error will be
+ * generated warning about multiple includes of setjmp (on some version of
+ * libpng). Because we don't use the setjmp functionality of freetype, this is
+ * not applicable in this case. Unfortunately, the compile error can only be
+ * fixed by editing png.h or reordering the includes.
+ *
+ * See https://sourceforge.net/p/enlightenment/mailman/message/12595025/ for
+ * some discussion on this topic.
+ */
+// clang-format off
+#include <png.h> // NOLINT
+#include <ft2build.h>
+#include FT_FREETYPE_H
+// clang-format on
+
+#include <llassetgen/Image.h>
+
+namespace llassetgen {
+
+	Image::Image(const FT_Bitmap &ft_bitmap) {
+		min.x = 0;
+		min.y = 0;
+		max.x = ft_bitmap.width;
+		max.y = ft_bitmap.rows;
+
+		switch (ft_bitmap.pixel_mode) {
+		case FT_PIXEL_MODE_MONO:
+			bit_depth = 1;
+			break;
+		case FT_PIXEL_MODE_GRAY2:
+			// We haven't found a single font using this format, however.
+			bit_depth = 2;
+			break;
+		case FT_PIXEL_MODE_GRAY4:
+			// We haven't found a single font using this format, however.
+			bit_depth = 4;
+			break;
+		default:
+			bit_depth = 8;
+		}
+
+		load(ft_bitmap);
+	}
+
+	void Image::load(const FT_Bitmap &ft_bitmap) {
+		assert(get_width() == ft_bitmap.width && get_height() == ft_bitmap.rows);
+
+		uint8_t ft_bit_depth;
+		switch (ft_bitmap.pixel_mode) {
+		case FT_PIXEL_MODE_MONO:
+			ft_bit_depth = 1;
+			break;
+		case FT_PIXEL_MODE_GRAY2:
+			// We haven't found a single font using this format, however.
+			ft_bit_depth = 2;
+			break;
+		case FT_PIXEL_MODE_GRAY4:
+			// We haven't found a single font using this format, however.
+			ft_bit_depth = 4;
+			break;
+		default:
+			ft_bit_depth = 8;
+		}
+		assert(ft_bit_depth == bit_depth);
+
+		stride = size_t(std::ceil(float(get_width() * bit_depth) / 8));
+
+		data = std::shared_ptr<uint8_t>(new uint8_t[get_height() * stride]);
+		for (size_t y = 0; y < get_height(); y++) {
+			for (size_t x = 0; x < stride; x++) {
+				data.get()[y * stride + x] = 0xFF - ft_bitmap.buffer[y * ft_bitmap.pitch + x];
+			}
+			size_t padding_bits = size_t((float(get_width() * bit_depth) / 8 - std::floor(float(get_width() * bit_depth) / 8)) * 8);
+			uint8_t mask = ~(0xFF >> padding_bits);
+			data.get()[y * stride + stride - 1] &= mask;
+		}
+	}
+
+	Image::Image(const size_t width, const size_t height, const size_t _bit_depth) {
+		assert(_bit_depth == 1 || _bit_depth == 2 || _bit_depth == 4 || _bit_depth == 8 || _bit_depth == 16 || _bit_depth == 24 || _bit_depth == 32);
+		min.x = 0;
+		min.y = 0;
+		max.x = width;
+		max.y = height;
+        bit_depth = _bit_depth;
+        stride = (size_t)std::ceil(float(get_width() * (float(bit_depth) / 8.0f)));
+		data = std::shared_ptr<uint8_t>(new uint8_t[get_height() * stride]());
+	}
+
+	size_t Image::get_width() const {
```

Comment:
I am not sure if we agreed on under_score for variable names. I thought we use camelCase?

get/set: if it's a getter, just name it width(). At least that's a naming convention that I follow a lot, but it's optional. We might discuss that in our project group, if you want.